Cookbook/Proposals/PropertyWrappers.md

```diff
@@ -0,0 +1,155 @@
+
+Before:
+
+```swift
+struct DemographicsListDTO: Decodable {
+  let demographics: [DemographicDTO]
+	
+  init(from decoder: Decoder) throws {
+    let values = try decoder.container(keyedBy: CodingKeys.self)
+    let demographics = try values.decodeArraySafely([DemographicDTO].self, forKey: .demographics)
+  }
+}
+```
+
+After:
+
+```swift
+struct DemographicsListDTO: Decodable {
+  @ArrayDecodable<Prune, DemographicDTO>
+  let demographics: [DemographicDTO]
+}
+```
+
+Providing defaults for absent values has a bit tricky ergonomics, as this approach does not allow to pass arbitrary parameters to property wrapper initialisers when they are being decoded (so `@Default(value: true)` is possible programmatically but won't affect decoding). So instead of using values we use types which would correspond to common defaults, i.e. `True`, `False`, `Empty` etc:
+
+```swift
+struct Product: Codable {
+  var name: String
+  
+  @DefaultDecodable<Empty>
+  var description: String
+  
+  @DefaultDecodable<True>
+  var isAvailable: Bool
+}
+```
+
+Reference: https://github.com/gonzalezreal/DefaultCodable
+
+Nested keys are a bit more tricky, but still possible:
+
+```swift
+struct InboxMessage: Decodable {
+  @NestedDecodable<Bool, IsReadKeys>
+  var isRead: Bool // { "status": { "is_read": true } }
+  
+  enum CodingKeys: String, CodingKey {
+    case isRead = "status"
+  }
+  
+  enum IsReadKeys: String, CodingKey, CaseIterable {
+    case isRead = "is_read"
+  }
+}
+```
+
+Here we specify additional coding keys type that will be used to decode `isRead`. Then container extension method takes care of traversing json using this additional keys and their order (via CaseIterable) as a key path.
+
+When needed we can provide Encoding and Codable wrappers, though in general we have either encodable or decodable types. We do have encoding implemented manually for types used by mock server in UI tests, so providing Codable wrappers for properties of these types would also decrease code in tests and will make it easier to add more types to mock server.
+
+
+## Impact on existing codebase
+
+To replace each helper methods property wrappers would be introduced via separate change to simplify transition and testing.
+
+
+## Alternatives considered
+
+1. We can come up we a single property wrapper to describe decoding strategy and provide the actual strategy as a generic parameter:
+
+```swift
+@Decoding<Default<Empty>>
+@Decoding<Array<Prune>>
+@Decoding<Nested<IsReadKeys>>
+```
+
+This might be a good way to unify approach, but it's a bit harder to read and it we would need to figure out how to constrain particular strategies to particular types of wrapped values, which seems to be easier with a more straight forward approach.
```

Comment:
```suggestion
This might be a good way to unify the approach, but it's a bit harder to read and we would need to figure out how to constrain particular strategies to particular types of wrapped values, which seems to be easier with a more straight forward approach.
```